WIP: Port of Xtreme Thinblocks and Xpress Validation

rusty-loy opened the pull request:
This PR is a work in progress, integrating the [BUIP010](https://bitco.in/forum/threads/buip010-passed-xtreme-thinblocks.774/) features Xtreme Thinblocks and Xpress Validation, the great work of Peter Tschipper(@ptschip), Andrew Stone(@gandrewstone), and the rest of the Bitcoin Unlimited team.  I will keep this PR up to date with the commits found in the [BUIP010](https://github.com/BitcoinUnlimited/BitcoinUnlimited/commits/BUIP010) branch.

I believe I have brought over all the relevant code for these features.  Everything is up to date, all tests pass, and the Classic Client exchanges xthinblocks with Bitcoin Unlimited clients.

Current Task List (which I'm sure will grow):
- [x]  Validate: Cross-reference changes made in Bitcoin Unlimited to ensure all modifications made for Xtreme TB and Xpress Validation are in this PR.
- [ ]  Test: Add unit testing, create integration testing.
- [ ]  Minimize commits: Squash commits into more logic groups.
- [ ]  Minimize code changes: Remove modifications not necessary for Xtreme TB and Xpress Validation.
- [ ]  Reduce statistics class size.
- [ ]  Fix compatibility with send headers.
- [ ]  Measure benefit of backup method ordinary thin blocks (possibly remove).
- [ ]  Simplify thin blocks: "send get_xthin instead of getdata if peer services includes NODE_XTHIN" and "send xthinblock instead of block if..."
- [ ]  Test: Add more unit testing, create integration testing.

Commit:
Xtreme Thinblocks

Sends a seeded bloom filter along with a getdata
and receives an xthinblock transaction in return which
includes all missing transactions and tx hashes.
Once received the block is reconstructed.

Additional bandwidth is saved by using a 64 bit
Tx hash instead of all 256 bits.  Hash collisions
in the memory pool will be extremely rare however
if one should happen then a regular thinblock will
be re-requested.

Transactions still missing will be re-requested and
retrieved from a block rather than the mempool. They
will be sent back in one thinblocktx object.

Originally Inspired by Mike Hearn
xthinblock + coinbase  (dagurval)

Conflicts:
	src/main.cpp
	src/net.h
	src/unlimited.cpp
	src/unlimited.h

Commit:
Xpress Validation

Building on Xtreme Thinblocks, Xpress Validation
allows for a much faster and more scalable approach
to validating a block.  Becuase most transactions have
already been validated when entering the memory pool
we only have to validate new trasactions arriving in the
XThinblock before re-assembling the block.

Conflicts:
	src/main.cpp

YarkoL committed:
thinblock unit tests

gandrewstone committed:
Fix issue testing if thin blocks is enabled -- instead of calling the function IsThinBlocksEnabled(), the address of it is tested against 0.

Fix locking order issue where cs_mapRelay is taken before cs_vSend, although most of the time the opposite is true.  This reverse order triggers the possible deadlock detector.

Fix signed/unsigned comparison complaints in leveldb.

gandrewstone committed:
small paren fix, BU mined blocks changed to indicate 2MB support

Conflicts:
	src/consensus/params.h
	src/miner.cpp

Commit:
Fix missing parenthesis in net.cpp for connect
and connect-thinblock functionality

Commit:
Fixed improperly formated LogPrint statement

Commit:
XThinblock preferential downloader

Thinblocks will be preferentially downloaded from nodes
that are connected and that support thinblocks.  If the
timer is exceeded then a regular block is downloaded instead.

Conflicts:
	src/unlimited.cpp

gandrewstone committed:
1. add node's IP address to thin block logs that reference a node to make the logs more understandable
2. Clear a CNode's thinblock CBlock structure via SetNull() rather than via copy constructor for clarity and efficiency.
3. Also clear the CNode's thinblock right away to ensure that is is not used after it has been "released"
4. In SendMessages, the deadlock detector is triggered, so reordered cs_main and cs_vSend locks to match other code's locking order (not a thin block issue)
5. Prefer thin block communications by handling them first in ThreadMessageHandler.
6. Add a member function ThinBlockCapable() that returns true if the node supports thin blocks

Commit:
Fix Windows Compile Error for Boost Interruption point

Commit:
Fix for locking issue and compile warning

1) Potential Deadlock using cs_filter:  (G.A. Stone)
2) Compile warning - no return value :  (sickpig)
3) Edit re-request thinblock message:   (sickpig)

gandrewstone committed:
1. add statistics that count the number of unnecessary full txs in a thin block
2. add statistics that calculate the ratio of full block size to thin block size
3. convert CXThinBlock::CXThinBlock filter to a pointer and send hashes only if its NULL
4. Add an (unused) CXThinBlock constructure that is hashes only
5. properly convert to new UniValue in unlimited CLI commands

Conflicts:
	src/unlimited.cpp

Commit:
Skip Xpress Validation if the thinblock previous block hash does not match the chaintip hash

Commit:
Created new NetMessage type GET_XTHIN

a GET_XTHIN message consists of an inv for an XTHINBLOCK
with a seeded bloom filter appended to it.  This prevents
the bloom filter from being processed at the other end
after the thinblock request, which can happen when the deque's
are backed up under heavy load.

hfinger committed:
fixed compression ratio reporting to include re-requested txs.

Commit:
Added LogPrint for size of seeded bloom filter

bitcartel committed:
Simple loop optimisation.

bitcartel committed:
Format whitespace and variable name.

Commit:
Orphan Cache default change and DB cache default change

Orphan Cache default to 5000 transactions.  Maximum orphan
tx size is 5000 bytes, therefore maximum cache size is now 25MB

Up the default db cache from 100MB to 300MB

Commit:
Xtreme Thinblocks and Xpress Validation using Orphan Cache

Many tx's now are found in the orphan cache for which there
is no parent tx.  When a thinblock is retrieved we have to
add the tx hashes to the bloom filter and also check the
orphan cache on return to see if there are any valid tx's to
add to the thinblock.  Furthermore we have to make sure we
do not invoke Xpress Validation for these orphan tx's as they
have never been fully verified and we must do so before accepting
the block.

Conflicts:
	src/main.cpp
	src/main.h

rusty-loy committed:
Add missing files from Bitcoin Unlimited necessary for Xtreme Thinblocks.

rusty-loy committed:
Remove unused code from unlimited.

Commit:
Added Service Bit #4 for Xtreme Thinblocks

Xtreme Thinblocks service enabled

NODE_XTHIN = (1 << 4),

Conflicts:
	src/main.cpp

Commit:
Optimize the messaging for XTHINBLOCK requests

No longer send a map of all tx's but instead
just send a vector.  This saves us 256bits per
transaction for XTHINBLOCK and THINBLOCK responses.

Commit:
Optimize GET_XTBLOCKTX and XBLOCKTX messages

No longer sending the 256 bit hash with these
messages.  Only sending a set or vector or tx's

rusty-loy committed:
Fix reject block message causing test p2p-fullblocktest to fail.

rusty-loy committed:
Disable thin blocks for sendheaders test.  TB does not support headers first.

bitcartel committed:
Add class CThinBlockStats to keep track of thin block statistics.

bitcartel committed:
Add thin block statistics to RPC call 'getnetworkinfo' under key
'thinblockstats'.

Commit:
Do not delete orphans from the cache on a node disconnect

For Xtreme Thinblocks we need to keep all orphans in the
cache so when a node disconnects we disable the funtionality
that causes the orphans that were propagated from that peer
to be purged from the cache.

Commit:
Update to tests to account for the change from map to vector

gandrewstone committed:
add statistics container and example instrumentation of nodes and the mempool

Conflicts:
	src/net.cpp
	src/net.h
	src/unlimited.cpp
	src/unlimited.h

rusty-loy committed:
Fix thinkblocks unit test, partial cherry-pick of 1d1cc7ef7588be9a7e303d6eb5c7de73598c074f.

hfinger committed:
fixed cheap hash collision detection

hfinger committed:
removed unnecessary logs

gandrewstone committed:
WIP: integration

Commit:
Added a separate bloom filter specifically for XThinblocks

Added pnode->pThinBlockFilter so as to keep us from
potentially having a conflict with pnode->pfilter which
is used by SPV wallets.  We don't want any potential
conflicts or have to worry about clearing or deleting
the filter after each use.

gandrewstone committed:
getstats: user convenience, if the number of items is not specified send just 1 not the whole history

gandrewstone committed:
Update to include code missed during previous merges.

gandrewstone committed:
oops, this filter lock is not really needed because filter won't be updated but creates a 'possible deadlock' with vSend

rusty-loy committed:
Cleanup from previous merges.

rusty-loy committed:
Fix txn_doublespend test.

Revert previous change found in commit b996db5fc1e6b1f09f32d99f6527191e89d6e267.

rusty-loy committed:
Fix unit test stat_tests.

keo commented:
:+1:

sandakersmann commented:
Concept ACK :+1:

zander commented:
We've been wanting to somehow integrate this concept :)

Referring to https://github.com/bitcoinclassic/bitcoinclassic/blob/0.12/CONTRIBUTING.md

I would suggest we follow the concept of a topic-branch where the only thing we look at for reviewing the PR is the final diff.

The "Topic" is important, though. You should try really hard to avoid changes that are irrelevant to the change that is being merged. I've commented on one of those changes that look not on-topic.

This sounds like something that would also not be on-topic and thus just make a review harder;

>  Refactor: Work with Bitcoin Unlimited team to see if we can minimize code in main.cpp (these merges are painful).

All in all, awesome stuff!

dagurval commented:
If you like I can review this PR when it's done. I'm also implementing xthin in BitcoinXT (https://github.com/bitcoinxt/bitcoinxt/pull/135) so I know the concept and the code well. Just notify me here or on slack.

gavinandresen commented:
I agree with Tom -- this should be a development branch that gets squashed, 42 commits is way too many to review.

I'm going to brain-dump comments as I review:

Changes to clean up leveldb compiler warnings should be submitted upstream, and should be a separate pull request.

600+ new lines of code for keeping some simple statistics seems excessive.  Why not use boost's statistics library? ( http://www.boost.org/doc/libs/1_60_0/doc/html/accumulators.html )

I don't see why this would be incompatible with "getheaders"/head-first. The protocol for node A that has a block sending to node B that wants it (both support thin blocks) should be:

```
A sends 'headers' or 'inv' to tell B there's a new block available
B replies with 'get_xthin" with the bloom filter instead of 'getdata'
A replies 'xthinblock' instead of 'block'
```

The comments in protocol.h should be improved to describe the thinblocks protocol (who sends what to who); comments like "The get_xthin message transmits a single serialized get_xthin." are useless.

The code for thin blocks feels like it needs a rewrite.  It really should be as simple as "send get_xthin instead of getdata if peer services includes NODE_XTHIN" and "send xthinblock instead of block if..." It looks like this code evolved from Mike's code (which was trying hard to be old-protocol-compatible).

Dropping back to 'ordinary thin blocks' in case the xthin thing doesn't work seems like a lot of extra code; can we measure the benefit? I suspect it'd be better to just drop back to getdata/block (since ordinary thin blocks might have to drop back to that method anyway).

The fixes to lock order warnings when compiled -DDEBUG_LOCKORDER should be a separate pull request (and I think implementing a REVERSE_LOCK() in sync.h might simplify at least one of those fixes quite a lot).

So: concept ACK, but I think the final pull request should be a fresh set of clean commits, with code copied from the great work the Unlimited team has done.

zander commented:
Agreed with everything Gavin said.  I have not done serious review since it was still marked as work-in-progress.  What is the status of the x-thin-blocks project?

If this can be squashed into a smaller amount of commits and the amount of code added can be minimized that would be really nice.

Just in case this sounds like something you didn't want to hear; I refer to the [Contributions](https://github.com/bitcoinclassic/bitcoinclassic/blob/develop/CONTRIBUTING.md#to-be-aware-of)  file which mentions things like atomic commits and similar ideas.

rusty-loy commented:
Thanks so much for the comments.  

The current status of the port is I just finished the first step of validating that I had pulled over all the code from https://github.com/BitcoinUnlimited/BitcoinUnlimited/commits/BUIP010.  All unit and integration tests pass and I have rebased to be caught up with the current dev branch.  Perhaps @gandrewstone  can comment on the status of the x-thin-blocks project?

@dagurval Thank you very much for the offer, I will definitely let you know when I am closer to a finishing state.  Your input is much appreciated.

@gavinandresen I have added some points in my task list based on your feedback.  I am sure we will also talk in slack as I have lots of questions!

@zander Will talk to you more on this.  I want to do the necessary clean up work that will make reviewing this easier for everyone.

ptschip commented:
Just to give you a little history of what was done at BU in regard to
xthins:

Originally we built it to work on 11.2...so there was no headers first
functionality there.  When 12.0 came along the header first was a
nuisance and not really necessary for what we were trying to do.  We
didn't know at the time that you were going to submit a PR on headers
first mining otherwise we might have persisted in working around the
issues there.  I can't speak for everyone at BU but I think you'd find a
great deal of support for your headers first mining and we would most
likely include any work you do to that end. 

Another piece of history is the functionality around when a hash
collision occurs and we re-request a thinblock rather than a full
block.  Originally we built the system around "thinblocks", it was going
to be called "Scalable Thinblocks", kind of a dull name but anyway
that's what we started with before we brought in the use of 64bit
hashes.  So we already had the thinblocks function and thought that we
should keep it because we really didn't know how many hash collisions
wer were going to have.  Theory said almost never, but practice is
sometimes different.  Well so far it looks like theory is doing well
BUT, that's for now running at 3 tx/second..What happens when we're at
100 tx/sec or more...will we start seeing collision there?  I don't
know, so we thought better to keep the function in rather than down the
road when we have 10 , 20, 30 MB blocks are we start getting hash
collisions and now downloading the full blocks and the p2p network is
having trouble.  I don't like that scenario or the possiblity of it.  I
think it's easier to sleep knowing that, that couldn't happen.  But of
course it's entirely up to you what you do, I just wanted to let you
know the thinking behind it was.

Also, we're moving pretty fast and putting out "good enough" software,
(hopefully a little better than good enough).   We've had a few other
newer developers submit some optimizations and bug fixes which is all
well and good also gets other developers in touch with the code base. 
That said, any further optimizing you do is welcome to us as well. 
Sorry for the lack of documentation but as I said we're moving fast and
we can always circle back later and fill in the gaps.

Yup, there are  lot of commits, we went through many iterations, release
cadidates, and bug fixes along the way.  You don't need those commits, I
don't see why you can't make one or two commits.  I think that's very
sensible.

On 24/03/2016 11:21 AM, Gavin Andresen wrote:

> I agree with Tom -- this should be a development branch that gets
> squashed, 42 commits is way too many to review.
> 
> I'm going to brain-dump comments as I review:
> 
> Changes to clean up leveldb compiler warnings should be submitted
> upstream, and should be a separate pull request.
> 
> 600+ new lines of code for keeping some simple statistics seems
> excessive. Why not use boost's statistics library? (
> http://www.boost.org/doc/libs/1_60_0/doc/html/accumulators.html )
> 
> I don't see why this would be incompatible with
> "getheaders"/head-first. The protocol for node A that has a block
> sending to node B that wants it (both support thin blocks) should be:
> 
> |A sends 'headers' or 'inv' to tell B there's a new block available B
> replies with 'get_xthin" with the bloom filter instead of 'getdata' A
> replies 'xthinblock' instead of 'block' |
> 
> The comments in protocol.h should be improved to describe the
> thinblocks protocol (who sends what to who); comments like "The
> get_xthin message transmits a single serialized get_xthin." are useless.
> 
> The code for thin blocks feels like it needs a rewrite. It really
> should be as simple as "send get_xthin instead of getdata if peer
> services includes NODE_XTHIN" and "send xthinblock instead of block
> if..." It looks like this code evolved from Mike's code (which was
> trying hard to be old-protocol-compatible).
> 
> Dropping back to 'ordinary thin blocks' in case the xthin thing
> doesn't work seems like a lot of extra code; can we measure the
> benefit? I suspect it'd be better to just drop back to getdata/block
> (since ordinary thin blocks might have to drop back to that method
> anyway).
> 
> The fixes to lock order warnings when compiled -DDEBUG_LOCKORDER
> should be a separate pull request (and I think implementing a
> REVERSE_LOCK() in sync.h might simplify at least one of those fixes
> quite a lot).
> 
> So: concept ACK, but I think the final pull request should be a fresh
> set of clean commits, with code copied from the great work the
> Unlimited team has done.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/bitcoinclassic/bitcoinclassic/pull/147#issuecomment-200958806

gandrewstone commented:
I think that if you pull into a branch and run "git merge --squashed" it'll
merge to your main branch in one commit while preserving attribution.

I can see why you may want to simplify.  We can easily add handling of a
"reject" message if you choose to not include thin blocks.  And yes there
is some forward-looking code whose value may not justify its complexity at
this moment... ultimately it is of course your choice what to include and I
am happy to respond to any questions about design choices or motivation, or
take a quick look at proposed changes.
On Mar 24, 2016 2:21 PM, "Gavin Andresen" notifications@github.com wrote:

> I agree with Tom -- this should be a development branch that gets
> squashed, 42 commits is way too many to review.
> 
> I'm going to brain-dump comments as I review:
> 
> Changes to clean up leveldb compiler warnings should be submitted
> upstream, and should be a separate pull request.
> 
> 600+ new lines of code for keeping some simple statistics seems excessive.
> Why not use boost's statistics library? (
> http://www.boost.org/doc/libs/1_60_0/doc/html/accumulators.html )
> 
> I don't see why this would be incompatible with "getheaders"/head-first.
> The protocol for node A that has a block sending to node B that wants it
> (both support thin blocks) should be:
> 
> A sends 'headers' or 'inv' to tell B there's a new block available
> B replies with 'get_xthin" with the bloom filter instead of 'getdata'
> A replies 'xthinblock' instead of 'block'
> 
> The comments in protocol.h should be improved to describe the thinblocks
> protocol (who sends what to who); comments like "The get_xthin message
> transmits a single serialized get_xthin." are useless.
> 
> The code for thin blocks feels like it needs a rewrite. It really should
> be as simple as "send get_xthin instead of getdata if peer services
> includes NODE_XTHIN" and "send xthinblock instead of block if..." It looks
> like this code evolved from Mike's code (which was trying hard to be
> old-protocol-compatible).
> 
> Dropping back to 'ordinary thin blocks' in case the xthin thing doesn't
> work seems like a lot of extra code; can we measure the benefit? I suspect
> it'd be better to just drop back to getdata/block (since ordinary thin
> blocks might have to drop back to that method anyway).
> 
> The fixes to lock order warnings when compiled -DDEBUG_LOCKORDER should be
> a separate pull request (and I think implementing a REVERSE_LOCK() in
> sync.h might simplify at least one of those fixes quite a lot).
> 
> So: concept ACK, but I think the final pull request should be a fresh set
> of clean commits, with code copied from the great work the Unlimited team
> has done.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/bitcoinclassic/bitcoinclassic/pull/147#issuecomment-200958806

zander commented:
Updated branch on https://github.com/zander/bitcoinclassic/tree/xtremetb

That one is up-to-date.

If anyone knows of changes where BU diverged from this initial list of cherry-picks, please let me know before I start reviewing that.

rebroad commented on src/main.cpp:
```diff
@@ -4134,20 +4173,20 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-            return false;
-        }
-    }
+// BUIP010 Extrem Thinblocks:  We need bloom filtering. We do not turn bloom filtering off
```
Bit drastic... perhaps only skip this code if the node in question is an XThin capable node?

zander replied:
You are commenting on very old code that has been changed quite a lot since.  This specific part has been re-added in a later commit by the guy that originally removed it ;)

If you want to review BitcoinClassic I suggest you look at the develop branch, not at a couple months old commits.

Notice that this port was marked as work-in-progress. It was worked on and the actual tree change is substantially smaller as a result.
